url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/cmdlib/BeanLibrary.java

```diff
@@ -0,0 +1,560 @@
+                {
+                    String lIdx = lIndices[i];
+                    int lClosing = lIdx.lastIndexOf(']');
+                    if(lClosing > 0) lIdx = lIdx.substring(0, lClosing).trim();
+                    lResult.add(lIdx);
+                }
+            }
+        }
+        return lResult;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void update(Object aTarget, Object aProp, Object aVal)
+            throws CommandException
+    {
+        String lPropName = "";
+        int lIdx = -1;
+
+        if(aProp instanceof String) lPropName = (String) aProp;
+        else throw new CommandException("Don't know what to do with non-string keys ... (for now).");
+
+        try
+        {
+            // Try to interprete the property as
+            // an integer.
+            lIdx = Integer.parseInt(lPropName);
+        }
+        catch (NumberFormatException e1)
+        {
+            // Ignore this exception!
+        }
+
+        if(aTarget instanceof List)
+        {
+            final List lTargetList = (List) aTarget;
+            if(lIdx < 0)
+                throw new CommandException("Modifying list should use an index.");
+            if(lIdx >= lTargetList.size())
+                throw new CommandException("Array index out of range.");
+
+            // TODO: catch errors on this one.
+            lTargetList.set(lIdx, aVal);
+
+        }
+        else if(aTarget != null)
+        {
+            if(aTarget.getClass().isArray())
+            {
+                // Array access.
+                if(lIdx < 0)
+                    throw new CommandException("Modifying array should use an index.");
+                if(lIdx >= Array.getLength(aTarget))
+                    throw new CommandException("Array index out of range.");
+
+                // TODO: catch errors on this one.
+                Array.set(aTarget, lIdx, aVal);
+            }
+            else
+            {
+                try
+                {
+                    // JavaBean access.
+                    BeanInfo lInfo = Introspector.getBeanInfo(aTarget.getClass());
+                    PropertyDescriptor lProps[] = lInfo.getPropertyDescriptors();
+                    for (PropertyDescriptor lProp : lProps)
+                    {
+                        if(lProp.getDisplayName().equals(lPropName))
+                        {
+                            Method lMeth = lProp.getWriteMethod();
+                            if(lMeth == null)
+                                throw new CommandException("Property cannot be written.");
+
+                            if(lProp.getPropertyType() != aVal.getClass() && aVal instanceof String)
+                            {
+                                PropertyEditor lEditor = lProp.createPropertyEditor(aTarget);
+                                lEditor.setAsText((String) aVal);
+                            }
+                            else lMeth.invoke(aTarget, new Object[]{aVal});
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)